util/src/MultiSizeBufferPool.cpp

```diff
@@ -0,0 +1,687 @@
+  waitForAvailChunkCond_.notify_all();
+
+  // check for a leak / inconsistency
+  if (stats_.current_.numUnusedChunks_ != 0) {
+    LOG_FATAL(logger_, "numUnusedChunks_ is non-zero!" << KVLOG(KVLOG_PREFIX) << stats_);
+    ConcordAssert(false);
+  }
+  if (stats_.current_.numAllocatedChunks_ != stats_.current_.numUnusedChunks_) {
+    auto kvlogArgs = KVLOG(stats_.current_.numAllocatedChunks_, stats_.current_.numUnusedChunks_);
+    if (poolConfig_.WarnOnSuspectedLeak) {
+      LOG_ERROR(logger_, "There is probably a memory leak! Not all chunks returned to subpool:" << kvlogArgs);
+    }
+  }
+  LOG_DEBUG(logger_, "Subpool destroyed:" << KVLOG(KVLOG_PREFIX, config_.bufferSize));
+}
+
+void MultiSizeBufferPool::SubPool::reportBufferUsage(char* chunk, uint32_t usageSize) {
+  auto* chunkHeader = reinterpret_cast<MultiSizeBufferPool::ChunkHeader*>(chunk);
+  DEBUG_PRINT(logger_, "Report buffer usage:" << KVLOG(KVLOG_PREFIX, chunkHeader->bufferSize_, usageSize));
+  if (usageSize > chunkHeader->bufferSize_) {
+    ++stats_.overall_.numBufferUsageFailed_;
+    ++poolStats_.overall_.numBufferUsageFailed_;
+  } else {
+    ++stats_.overall_.numBufferUsageSuccess_;
+    ++poolStats_.overall_.numBufferUsageSuccess_;
+  }
+}
+
+void MultiSizeBufferPool::SubPool::deleteChunk(char*& chunk) {
+  DEBUG_PRINT(logger_, "deleteChunk enter:" << KVLOG(KVLOG_PREFIX, (void*)chunk));
+  delete[] chunk;
+  stats_.onChunkDeleted(chunkSize_);
+  poolStats_.onChunkDeleted(chunkSize_);
+
+  DEBUG_PRINT(logger_, "deleteChunk exit:" << KVLOG(KVLOG_PREFIX, (void*)chunk) << stats_);
+  chunk = nullptr;
+}
+
+MultiSizeBufferPool::SubPool::AllocationResult MultiSizeBufferPool::SubPool::getChunk(bool wait, bool allocate) {
+  AllocationResult result{nullptr, AllocationStatus::SUCCESS};
+  auto& chunk = result.first;
+  bool chunkAllocatedFromHeap{false};
+
+  DEBUG_PRINT(logger_, "getChunk enter" << KVLOG(KVLOG_PREFIX));
+
+  if (!chunks_.pop(chunk)) {
+    // No available chunks => allocate a new one from the heap, if permitted
+    if (allocate) {
+      result = allocateChunk();
+      if (result.second == AllocationStatus::EXCEED_POOL_MAX_BYTES) {
+        DEBUG_PRINT(logger_, "getChunk exit" << KVLOG(KVLOG_PREFIX));
+        return result;
+      }
+      chunkAllocatedFromHeap = (chunk != nullptr);
+    }
+    if (!chunk && wait) {
+      // Pool size limit has been reached: wait until some chunk gets released. The wait is done with sampling timeout,
+      // in case  stopFlag_ is raised, and thread needs to exit without a chink since pool is stopped.
+      unique_lock<mutex> lock(waitForAvailChunkMutex_);
+      bool isChunkAvailable{false};
+      const auto waitFunc = [&, this]() {
+        isChunkAvailable = chunks_.pop(chunk);
+        return isChunkAvailable || stopFlag_;
+      };
+
+      DEBUG_PRINT(logger_, "Wait for chunk:" << KVLOG(KVLOG_PREFIX));
+      waitForAvailChunkCond_.wait(lock, waitFunc);
```

Comment:
I'm not sure this code works correctly in case we enter the destructor in parallel. Probably you need to signal waitForAvailChunkCond_ first in the destructor to cause it to call the waitFunc function.